Add unit tests for COCOReader

awolant opened the pull request:
Add unit testing for various options added to `COCOReader` during SSD debugging effort. It uses small COCO-like dataset added to DALI_extra.

Signed-off-by: Albert Wolant <awolant@nvidia.com>

Kh4L commented:
Build 688417

kychennv commented:
could you please tell me what is the purpose of this test?  it is very glad that you add some comments for it. such like the purpose, the expect result

awolant commented:
During the debugging of SSD support we added couple of options to `COCOreader`: `shuffle_after_epoch`, `skip_empty` and `size_threshold`. Also `stick_to_shard` was added. We did not have unit tests for all of those and for `COCOReader` in general. I added them as a part of an effort to improve detection pipeline test coverage. As those are unit tests, expected result for each of them is described in `Asserts`.

awolant commented:
Rebuilds 688934

jantonguirao commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,302 @@
+    MemCopy(
+      boxes.data(),
+      boxes_output.data<float>(),
+      ObjectCount() * 4 * sizeof(float));
```
nitpick. Make 4 a variable

awolant replied:
Done

jantonguirao commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,302 @@
+  auto ids_epoch_2 = this->CopyIds(ws);
+
+  for (int id = 0; id < this->SmallCocoSize(); ++id) {
+    if (ids_epoch_1[id] != ids_epoch_2[id])
```
`ASSERT_EQ(ids_epoch_1[id], ids_epoch_2[id]);`
and remove the `FAIL()` down below

awolant replied:
Actually, this test is supposed to check, that `ids_epoch_1` and `ids_epoch_2` are not equal on at least one element (they should be the same array shuffled in different ways). So neither `ASSERT_EQ` nor `ASSERT_FALSE(a == b)`  don't really fit here.
So what I did there was something like: go through both vectors and compare elements. When ypu will find difference thetn `return`, as this is a success. When you reached the end of both vectors and there were no differences, this is `FAIL`

jantonguirao replied:
can you set a bool and then assert that the bool is your expected value? I don't really like that FAIL call

awolant replied:
Done

szalpal commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,301 @@
+
+  DeviceWorkspace ws;
+  pipe.RunCPU();
+  pipe.RunGPU();
```
Isn't `RunMixed()` required here? @klecki ?

awolant replied:
`RunMixed` is no more there

szalpal commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,301 @@
+  int SmallCocoSize() { return 64; }
+  int EmptyImages() { return 2; }
+  int ImagesWithBigObjects() { return 2; }
+  int ObjectCount() { return 194; }
```
How about making these functions `constexpr` (and free)?

awolant replied:
I don't know about it. It's not like they have any meaning without context of this class. They are directly connected with `CocoReaderTest` fields. And you can't make them `constexpr` as members, so I think I will leave this as is.

szalpal commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,301 @@
+namespace dali {
+
+class CocoReaderTest : public ::testing::Test {
+ public:
```
```suggestion
 protected:
```

awolant replied:
Done

szalpal commented on dali/pipeline/operators/reader/coco_reader_op_test.cc:
```diff
@@ -0,0 +1,301 @@
+    MemCopy(
+      ids.data(),
+      output.data<int>(),
+      shape.size() * sizeof(int));
```
I'm not 100% sure here, but `sizeof(int)` seems to be unsafe, cause it's size depends on architecture. Might this be a problem?

Kh4L replied:
Should be fine, since the data type is also `int`
